src/main/java/hudson/plugins/ec2/util/MinimumInstanceChecker.java

```diff
@@ -74,8 +75,46 @@ public static int countQueueItemsForAgentTemplate(@NonNull SlaveTemplate agentTe
                 .count();
     }
 
-    public static void checkForMinimumInstances() {
-        Jenkins.get().clouds.stream()
+    /**
+     * JENKINS-76171: This method is synchronized to prevent race conditions during concurrent invocations.
+     *
+     * PROBLEM: This method is called from EC2RetentionStrategy.taskAccepted() when an agent
+     * accepts a task. With maxTotalUses=1, multiple agents can accept tasks simultaneously,
+     * causing multiple threads to call this method concurrently. Without synchronization,
+     * each thread independently calculates capacity and provisions nodes, leading to massive
+     * over-provisioning (e.g., 500+ nodes for 100 builds).
+     *
+     * SOLUTION: The synchronized keyword ensures only one thread executes this method at a time.
+     * When thread A provisions nodes, threads B, C, D wait. When thread B enters, it sees the
+     * nodes thread A just provisioned and provisions fewer/none, preventing duplicates.
+     *
+     * PERFORMANCE: Early-exit optimization ensures minimal synchronized block duration when
+     * minimumNumberOfInstances and minimumNumberOfSpareInstances are both 0 (recommended).
```

Comment:
This category of comment seems like it belongs in the PR rather than permanently in the code. Code comments should cover what the code does _now_ and why (when it is not obvious from reading).